test/test_ourdna_dashboard.py

```diff
@@ -0,0 +1,397 @@
+from collections import OrderedDict
+from datetime import datetime
+from math import ceil
+from test.testbase import DbIsolatedTest, run_as_sync
+
+from db.python.layers.ourdna.dashboard import OurDnaDashboardLayer
+from db.python.layers.participant import ParticipantLayer
+from db.python.layers.sample import SampleLayer
+from db.python.tables.sample import SampleFilter
+from db.python.utils import GenericFilter
+from models.models import (
+    ParticipantUpsert,
+    ParticipantUpsertInternal,
+    SampleUpsert,
+    SampleUpsertInternal,
+)
+
+
+def str_to_datetime(timestamp_str):
+    """Convert string timestamp to datetime"""
+    return datetime.strptime(timestamp_str, '%Y-%m-%d %H:%M:%S')
+
+
+class OurDNADashboardTest(DbIsolatedTest):
+    """Test sample class"""
+
+    @run_as_sync
+    async def setUp(self) -> None:
+        # don't need to await because it's tagged @run_as_sync
+        super().setUp()
+        self.odd = OurDnaDashboardLayer(self.connection)
+        self.sl = SampleLayer(self.connection)
+        self.pl = ParticipantLayer(self.connection)
+
+        participants = await self.pl.upsert_participants(
+            [
+                ParticipantUpsertInternal(
+                    external_id='EX01',
+                    reported_sex=2,
+                    karyotype='XX',
+                    meta={'consent': True, 'field': 1},
+                ),
+                ParticipantUpsertInternal(
+                    external_id='EX02',
+                    reported_sex=1,
+                    karyotype='XY',
+                    meta={'field': 2},
+                ),
+                ParticipantUpsertInternal(
+                    external_id='EX03',
+                    reported_sex=2,
+                    karyotype='XX',
+                    meta={'consent': True, 'field': 3},
+                ),
+            ],
+        )
+
+        self.participants_external_objects = [participant.to_external() for participant in participants]
+
+        samples_data = [
```

Comment:
Why not construct the objects here directly rather than matching keys you need to select for, and instead keep them nested in the `ParticipantUpsert` (under the samples attribute) that will do all this inserting for you, eg: https://github.com/populationgenomics/metamist/blob/392d08191a79fbdeb2ac10f6a36c058111a007dc/test/test_graphql.py#L27-L31